### PR TITLE
HIVE-28480: Disable SMB on partition hash generator mismatch across join branches…

### PR DESCRIPTION
**Pull request by himanshu-mishra**
… in previous RS

### What changes were proposed in this pull request?

As SMB replaces last RS op from the joining branches and the JOIN op with MERGEJOIN, we need to ensure the RS before these RS, in both branches, are partitioning using same hash generator.

Hash code generator differs based on ReducerTraits.UNIFORM i.e. ReduceSinkOperator#computeMurmurHash()  or ReduceSinkOperator#computeHashCode(), leading to different hash code for same value.

Skip SMB join in such cases.

### Why are the changes needed?
It fixes bug that results in incorrect join result. Example query and behavior explanation is added in https://issues.apache.org/jira/browse/HIVE-28480

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
Tested via `auto_sortmerge_join_18.q` added in this PR

### CODE REVIEW

**Comment by ngsg** [comment]
The patch seems reasonable. LGTM

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java:882]: Can I assume we want to find ReduceSinkOperator-1 of `Operators -> ReduceSinkOperator-1 -> Operators -> ReduceSinkOperator-2(parentOp) ->JoinOperator`?

**Review by okumin — COMMENTED** [review]
This looks fine to me at this point.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java
Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java:882]: That makes sense. Thanks

**Comment by himanshu-mishra** [comment]
@okumin if this looks good, please merge the PR

**Comment by okumin** [comment]
@himanshu-mishra Sorry, I don't have the right to merge a PR as I am not a committer. You need approval of any committer first.

**Review by okumin — APPROVED** [review]

**Comment by himanshu-mishra** [comment]
@kasakrisz can you please review or add reviewer

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java
Comment (kasakrisz) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java:889]: There are several utility methods to find operators in branches of a plan in `ql.exec.OperatorUtils`
```
OperatorUtils.find*Operator*()
```
Is it possible to use one of them to find parent RS operators?

**Review by kasakrisz — COMMENTED** [review]
This is a good candidate to fix this issue, but have you checked why the reducer traits are different?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java
Comment (himanshu-mishra) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java:889]: There isn't a utility method there that returns first ancestor operator encountered from given set of classes. Here, we are only interested in ancestor RS if that is the source of data for current RS, if TS, JOIN are encountered, they are the source, and we do not need verifying them.

**Comment by himanshu-mishra** [comment]
> This is a good candidate to fix this issue, but have you checked why the reducer traits are different?

Consider the sample query
```sql
SELECT * FROM (
    SELECT k, COUNT(DISTINCT v), SUM(v)
    FROM t_asj_18 GROUP BY k
) a LEFT JOIN (
    SELECT k, COUNT(v)
    FROM t_asj_18 GROUP BY k
) b ON a.k = b.k; 
```

In this query the aggregation `k, count(distinct v).. group by k` leads to group by and hence following RS operator's keys to be `k, v` while keeping partition column `k`, as we are aggregating on `k`. In such cases where key columns and partition columns differ, the `UNIFORM` trait is removed [refer commit](https://github.com/apache/hive/commit/d0acc5d309bba3bb4e6110f92d10e1ea247edd7e).
Note that the group by keys and partition columns differ only with `distinct` aggregate, removing that would lead to both being `a` thereby having `UNIFORM` trait.

**Comment by himanshu-mishra** [comment]
@kasakrisz let me know if there are pending concerns.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java:889]: a few lines earlier the same while loop exists https://github.com/apache/hive/blob/b13f529d944dc4c72d02d18e63757622985ada28/ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java#L853-L859

Could you please extract these to a utility method and move it to `OperatorUtils`

**Comment by kasakrisz** [comment]
> @kasakrisz let me know if there are pending concerns.

Sorry, I was occupied in the last few days. I left only one comment.

**Review by kasakrisz — APPROVED** [review]
+1

**Pull request merged**
